Add ability to use APCu for translation cache

kynx opened the pull request:
First off, thanks for this library! It's just what I need as I move my current project to Alpine 😀 

To make things even faster, I've added the ability to use APCu for caching the translations. This drastically improves performance and reduces memory usage with large translation files:

```
➜  motranslator git:(apcu-cache) ✗ php benchmark.php 
Execution took 17.254643917084 seconds
➜  motranslator git:(apcu-cache) ✗ php benchmark-apcu.php
Execution took 0.032136917114258 seconds
```

There is a slight performance hit when using in-memory cache, compared to the original code - around 0.0006s per request. I'm hoping that's acceptable:

```
➜  motranslator git:(master) ✗ php benchmark.php
Execution took 14.997493028641 seconds
```

I'm getting one PHPStan fail locally. That's not in code I touched, and actually looks like it's caused by a workaround to a previous false positive. I'll see if GH gets the same and if so fix it in a separate commit.

williamdes commented on src/Cache/ApcuCache.php:
```diff
@@ -0,0 +1,127 @@
+        bool $reloadOnMiss = true,
+        string $prefix = 'mo_'
+    ) {
+        // @codeCoverageIgnoreStart
```
I think we can cover this too, by adding two different matrix one without and one with the extension and also writing a test for it

kynx replied:
Yup, didn't think of that. I've added the test and the matrix setup. I _think_ I did that last bit right... 😉

williamdes commented on .github/workflows/tests.yml:
```diff
@@ -26,7 +26,8 @@ jobs:
               with:
                   php-version: ${{ matrix.php-version }}
-                  extensions: mbstring
+                  extensions: apcu, mbstring
+                  ini_values: apc.enable_cli=1
```
Here is the place we could matrix this to cover with and without the extension

williamdes reviewed (approved):
This is so much excellent code, I have barely nothing to say 🏆

williamdes commented:
For some odd reason PHP 7.1 fails: https://github.com/phpmyadmin/motranslator/runs/5361065506?check_suite_focus=true
Probably somewhat related to phpstan: https://github.com/phpmyadmin/motranslator/runs/5361065508?check_suite_focus=true#step:6:17

kynx commented:
Thanks for the quick review @williamdes ! I've updated the tests for PHP7.1 and phpunit-7.5 plus added a couple more to cover the thrown exceptions.

williamdes commented on .github/workflows/tests.yml:
```diff
@@ -26,7 +27,8 @@ jobs:
               with:
                   php-version: ${{ matrix.php-version }}
-                  extensions: mbstring
+                  extensions: apcu, mbstring
```
maybe we could also not install the extension to be sure there is no possible side effects on regular user installations

williamdes replied:
https://docs.github.com/en/actions/learn-github-actions/expressions

kynx replied:
OK, I think that's done. I've also removed `ext-apcu` from `require-dev`.

ibennetch reviewed (approved):
I think this is great. Excellent work and wonderful job on the documentation as well.

I would also like to ensure the tests cover both APCu and non-APCu scenarios, but other than that I have nothing to add to what you've done and what William's already said.

Thanks for contributing this back to the project.

williamdes reviewed (approved):
This is perfect now
I think we need to do a patch release before merging this in order not to block the symfony version bump
Because this will require a new major version if I am not mistaken as the code is not backwards compatible

williamdes commented:
@kynx so you know why some lines are not covered?

ibennetch commented:
Ok, assigning this to me for a bit so I can do the next release before the merge.

kynx commented:
@williamdes I’ll check the coverage in the morning. I think I need another test or two to ensure there isn’t a race in that `reloadOnMiss` stuff too - it went in late and another look at it with a fresh head would be good :|

williamdes commented:
I was asking myself how we would be sure that when the user updates its phpmyadmin installation the mo translations would update?
Using a custom key prefix?

kynx commented:
@ibennetch I've added tests for the uncovered lines and fixed a race. Think it's good to go now 😄 

@williamdes I've added a bit to the README about reloading the translations on the fly. Is that what you were looking for?

Edit: Grrr... that test for apcu not installed isn't working and I can't figure it out: https://github.com/phpmyadmin/motranslator/runs/5375702108?check_suite_focus=true

It's only installed `mbstring` - is it possible `apcu` is installed in the base image?

williamdes commented:
> @ibennetch I've added tests for the uncovered lines and fixed a race. Think it's good to go now smile
> 
> @williamdes I've added a bit to the README about reloading the translations on the fly. Is that what you were looking for?

Kind of, but thanks for adding this too !
I think we will do something using "custom prefix for cache entries" to enable multiple versions of phpMyAdmin running and also to have version independent cache entries.
Probably something like `phpMyAdmin_{version}_mo_`

For me this PR is ready to be merged and shipped

williamdes commented on tests/Cache/ApcuDisabledTest.php:
```diff
@@ -0,0 +1,26 @@
+            $this->markTestSkipped('ext-apcu is enabled');
+        }
+
+        $this->expectException(CacheException::class);
```
```suggestion
        $this->expectException(CacheException::class);
        $this->expectExceptionMessage('ACPu extension must be installed and enabled');
```

williamdes commented on .github/workflows/tests.yml:
```diff
@@ -15,6 +15,7 @@ jobs:
         strategy:
             matrix:
                 php-version: ["7.1", "7.2", "7.3", "7.4", "8.0", "8.1"]
+                extensions: ["none, mbstring", "none, apcu, mbstring"]
```
You will need to add more extensions, see:
https://github.com/phpmyadmin/sql-parser/blob/0147fc9a96228156adf0ffea7772ce7c47a20c22/.github/workflows/tests.yml#L37

kynx replied:
Yeah. Working out all that it needs looked hard - I've just changed it to `[":apcu, mbstring", "apcu, mbstring"]`

williamdes replied:
Perfect, thank you for the attention you did bring on this concern

williamdes commented on .github/workflows/tests.yml:
```diff
@@ -15,6 +15,7 @@ jobs:
         strategy:
             matrix:
                 php-version: ["7.1", "7.2", "7.3", "7.4", "8.0", "8.1"]
+                extensions: [":apcu, mbstring", "apcu, mbstring"]
```
```suggestion
                # ":" disables the extension (https://github.com/marketplace/actions/setup-php-action#extensions-optional)
                extensions: [":apcu, mbstring", "apcu, mbstring"]
```

williamdes commented:
> First off, thanks for this library! It's just what I need as I move my current project to Alpine

Thank you for using and improving this library ;)
Would you mind sharing the link if it is public ?

kynx commented:
Sorry to disappear - work got hectic. I've just pushed a fix that came out in testing.

williamdes commented:
> Sorry to disappear - work got hectic. I've just pushed a fix that came out in testing.

No worries ;)
Thank you for the fix and the test

MauricioFauth commented on src/Translator.php:
```diff
@@ -101,77 +102,12 @@ class Translator
-     * @param string|null $filename Name of mo file to load (null to not load a file)
-     */
-    public function __construct(?string $filename)
+    public function __construct(CacheInterface $cache)
```
What about still accepting the filename to avoid a BC break and only accept CacheInterface on next major?

```suggestion
    /**
     * @param CacheInterface|string|null $filenameOrCache
     */
    public function __construct($filenameOrCache)
```

williamdes replied:
That would definitely help anyone using or packaging this library

williamdes replied:
If zero BC breaks can be made the I would be grateful for that
It would allow me to package the new version in Debian without having to wait

kynx replied:
@williamdes I've updated the constructor to accept `CacheInterface|string|null`

williamdes commented on src/Translator.php:
```diff
@@ -101,77 +103,19 @@ class Translator
 
     /**
-     * @param string|null $filename Name of mo file to load (null to not load a file)
+     * @param CacheInterface|string|null $cache
```
```suggestion
     * @param CacheInterface|string|null $cache Name of mo file to load (null to not load a file) or a CacheInterface implementation
```

williamdes reviewed (approved):
Thank you again
So this has no BC breaking changes, we may just merge this and make a release @mauriciofauth ?
